### ISSUE TITLE
Infinite loop for empty Markdown file

### ISSUE THREAD

**Issue by gkm2164** [planet42/Laika#89]
Hi, guys, I've found some issue.

The problem is that when I put 0 byte file, it falls into infinite loop.
Here's the code fragment.

```scala
Transform.from(Markdown).to(HTML).fromFile("some-file-which-is-0-byte.md")
```

I know you might probably busy but, could you check this for a moment?

Thanks.

**Comment by gkm2164** [comment]
Well I think the problem is in following code.
at 0.10.0 version
Location: Laika.parse
object ParserContext (111 ~ 122 line)

```
def apply (input: java.io.Reader, sizeHint: Int): ParserContext = {

    val arr = new Array[Char](sizeHint)
    val buffer = new StringBuilder
    var numCharsRead: Int = 0

    while ({numCharsRead = input.read(arr, 0, arr.length); numCharsRead != -1}) {
      buffer.appendAll(arr, 0, numCharsRead)
    }

    apply(buffer.toString)
  }
```

Well, at the while loop (when the file size is 0 byte) , ```numCharsRead``` seems to throw numCharsRead == 0, and since numCharsRead checks only for value ```-1``` in this case, so it falls into infinite loop.
But I think at newer version of this code, this part is deleted somehow, that is, it might be solved in the current master branch...

So, when will you gonna release the current version?

**Comment by jenshalm** [comment]
Hello and thanks for reporting.

You were pretty close to finding the cause, but the root problem is here: https://github.com/planet42/Laika/blob/0.11.0/core/src/main/scala/laika/io/Input.scala#L123
For an empty file it will create a sizeHint of zero, which in the code you posted above will lead to the creation of an Array with 0 length, so trying to read into it will never reach the end of file.

The loop on the other hand must continue for 0, this is standard Java stream I/O. Only a -1 marks the end of input.

Regarding releases, the next one is 0.11.0 (adding EPUB support), but that is not on master, it's on a branch named `0.11.0` (my link above points to that branch). That version is completed and ready to go, but it's currently parked as I wait for the final 2.13 release. Once that is out (looks like it's gonna happen in the coming days) and a corresponding release of ScalaTest I will publish Laika 0.11.0. I will add the fix for this as it's only a one line fix.

The master branch you browsed is currently in a state of flux, as 0.12.0 will come with a big refactoring and integration with cats-effect. For that the entire file I/O part will be completely rewritten (to be referentially transparent). I would recommend to ignore master for the coming 5 weeks or so as it is not in a stable state and many of the changes are not documented yet. The current estimate for a 0.12 release is sometime in August.

**Comment by jenshalm** [comment]
Oh, and just out of curiosity: why do you work with empty files? They don't seem to be particularly useful. :-)

**Comment by gkm2164** [comment]
@jenshalm 
First of all, thank you for your detailed explanation about the issue.
So the conclusion is that it would take some time for the next release.
Then, since I need to solve this problem, I should check whether the file size is 0 or not.
I will stare this library :)

For the question below, 
Well, I'm using your library on developing a generator which takes inputs as source codes and produces HTML document(it is similar with java-doc with some custom features)
But I wanted to add importing introduction markdown which is stored in the root of the source code,  and convert it to HTML and paste them on the top of the document.
But sometimes, well, there might be some occasions that developer who uses my tool, could generates those markdown file without any contents.(I mean, a file with 0 byte)
Since the introduction file is not necessary in my case, I need to allow those files.

**Comment by jenshalm** [comment]
No, the next release will happen pretty soon, sorry if I wasn't clear. I just explained that the next release is on the branch 0.11.0 and not on master. 0.11 will happen in the next week or the week after, as soon as Scala 2.13 is out.

The release which is farther away is the one coming from master (0.12.0) which will happen around August, but that's not the next release.

**Comment by gkm2164** [comment]
@jenshalm 
Ah, I see. I've been misunderstood.
Thanks!

**Comment by jenshalm** [comment]
fixed in both, the 0.11.0 branch and master (the fixes are in different places as master has a completely redesigned file I/O). I will now prepare the 0.11.0 release as Scala 2.13 is finally out!
